src/store/index-level.ts

```diff
@@ -1,213 +1,284 @@
+
+
+type IndexedItem<T> = {
+  value: T;
+  key: string;
+};
+
+const INDEX_SUBLEVEL_NAME = 'index';
 
 export interface IndexLevelOptions {
   signal?: AbortSignal;
 }
 
 /**
- * A LevelDB implementation for indexing the messages stored in the DWN.
+ * A LevelDB implementation for indexing the messages and events stored in the DWN.
  */
-export class IndexLevel {
-  config: IndexLevelConfig;
-
-  db: LevelWrapper<string>;
-
-  constructor(config: IndexLevelConfig) {
-    this.config = {
-      createLevelDatabase,
-      ...config
-    };
-
-    this.db = new LevelWrapper<string>({ ...this.config, valueEncoding: 'utf8' });
-  }
-
-  async open(): Promise<void> {
-    return this.db.open();
-  }
-
-  async close(): Promise<void> {
-    return this.db.close();
-  }
+export class IndexLevel<T> {
+  constructor(private db: LevelWrapper<string>) {}
 
   /**
-   * Adds indexes for a specific data/object/content.
-   * @param dataId ID of the data/object/content being indexed.
+   * Put an item into the index using information that will allow it to be queried for.
+   *
+   * @param tenant
+   * @param key a unique key that represents the item being indexed, this is also used as the cursor value in a query.
+   * @param value the value representing the data being indexed.
+   * @param indexes - (key-value pairs) to be included as part of indexing this item. Must include at least one indexing property.
+   * @param sortIndexes - (key-value pairs) to be used for sorting the index. Must include at least one sorting property.
+   * @param options IndexLevelOptions that include an AbortSignal.
    */
   async put(
     tenant: string,
-    dataId: string,
-    indexes: { [property: string]: unknown },
+    key: string,
+    value: T,
+    indexes: { [key:string]: unknown },
+    sortIndexes: { [key:string]: unknown },
     options?: IndexLevelOptions
   ): Promise<void> {
-    const partition = await executeUnlessAborted(this.db.partition(tenant), options?.signal);
+    // ensure sorted indexes are flat and exist
+    sortIndexes = flatten(sortIndexes);
+    if (!sortIndexes || Object.keys(sortIndexes).length === 0) {
+      throw new Error('must include at least one sorted index');
+    }
+
+    // ensure indexable properties exist
     indexes = flatten(indexes);
+    if (!indexes || Object.keys(indexes).length === 0) {
+      throw new Error('must include at least one indexable property');
+    }
 
-    const operations: LevelWrapperBatchOperation<string>[] = [ ];
+    const indexOps: LevelWrapperBatchOperation<string>[] = [];
+    // adding a reverse lookup to be able to delete index data as well as look up sorted indexes by a cursor
+    indexOps.push({ type: 'put', key: `__${key}__indexes`, value: JSON.stringify({ indexes, sortIndexes }) });
+
+    // for each indexable property we go through the different sorting indexes and construct a sorted index key.
+    // the sort property is the last property in the key before the tie-breaker.
```

Comment:
I'm not sure I understand the question.

However I will modify the comment, as I was reading the entire thing I realized that there is some confusion in the way that it is worded.

```suggestion
    // since LevelDB keys are sorted lexicographically, we need to create sortable keys for each indexable property.
    // for each indexable property we go through the different sorting indexes and construct a unique sorted index key.
    // the sort property is the last property in the key before the tie-breaker, the key itself is used as a tie-breaker.
```